test/framework/src/main/java/org/elasticsearch/cluster/service/ClusterStateTaskExecutorUtils.java

```diff
@@ -123,6 +123,7 @@ public void success(Runnable onPublishSuccess, ClusterStateAckListener clusterSt
             assert clusterStateAckListener != null;
             assert task == clusterStateAckListener || (task instanceof ClusterStateAckListener == false);
             this.succeeded = true;
+            onPublishSuccess.run();
```

Comment:
Ah no we shouldn't run `onPublishSuccess` here, we're still executing the update at this point. We should hold on to the success callback and run it once the update is fully computed and "published" (whatever that might mean in these tests).

Reply:
But there's nothing actually to do here, so the update is trivially complete anyway?

Reply:
Hm that shouldn't be the case always, we might be simulating running a batch of tasks, the first of which succeeds (and gets here) and then the second throws an exception, aborting the whole batch.

Reply:
Ah right, I've added a capture-and-call mechanism instead